### PR TITLE
eth/abi: support dynamic array encoding

### PR DESCRIPTION
**Pull request by peter-chung-xfers**
- support dynamic array encoding
- fix an issue that function signature calculating might be incorrect when there are params in type of dynamic array

### CODE REVIEW

File: spec/eth/abi_spec.rb [hunk 1]
```diff
@@ -9,6 +9,31 @@
     let(:basic_abi_tests_file) { File.read "spec/fixtures/ethereum/tests/ABITests/basic_abi_tests.json" }
     subject(:basic_abi_tests) { JSON.parse basic_abi_tests_file }
 
+    describe "dynamic encode" do
+      it "can encode array of string" do
+        encoded = Eth::Util.bin_to_hex(described_class.encode(["string[]"], [["hello", "world"]]))
+        expected = "0000000000000000000000000000000000000000000000000000000000000020000000000000000000000000000000000000000000000000000000000000000200000000000000000000000000000000000000000000000000000000000000400000000000000000000000000000000000000000000000000000000000000080000000000000000000000000000000000000000000000000000000000000000568656c6c6f0000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000005776f726c64000000000000000000000000000000000000000000000000000000"
+        pp encoded
```
Comment (q9f) [inline spec/eth/abi_spec.rb:16]:
```suggestion
```

**Review by q9f — APPROVED** [review]

**Comment by q9f** [comment]
Thank you! Could you run `rufo` and remove the stray `pp`?

**Comment by q9f** [comment]
Thank you, can you fix the tests?

**Comment by peter-chung-xfers** [comment]
> Thank you, can you fix the tests?

fixed

**Review by q9f — APPROVED** [review]

**Comment by peter-chung-xfers** [comment]
Hey @q9f just wondering if you have found any viable testsuite for this?

**Comment by q9f** [comment]
No, but there is an open ticket here: https://github.com/ethereum/tests/issues/691

**Pull request merged**